plugins/hls-overloaded-record-dot-plugin/src/Ide/Plugin/OverloadedRecordDot.hs

```diff
@@ -0,0 +1,240 @@
+import           Development.IDE.Spans.Pragmas  (NextPragmaInfo (..),
+                                                 getFirstPragma,
+                                                 insertNewPragma)
+import           Development.IDE.Types.Logger   (Priority (..), cmapWithPrio,
+                                                 logWith, (<+>))
+import           GHC.Generics                   (Generic)
+import           Ide.Plugin.RangeMap            (RangeMap)
+import qualified Ide.Plugin.RangeMap            as RangeMap
+import           Ide.PluginUtils                (getNormalizedFilePath,
+                                                 handleMaybeM, pluginResponse)
+import           Ide.Types                      (PluginDescriptor (..),
+                                                 PluginId (..),
+                                                 PluginMethodHandler,
+                                                 defaultPluginDescriptor,
+                                                 mkPluginHandler)
+import           Language.LSP.Types             (CodeAction (..),
+                                                 CodeActionKind (CodeActionRefactorRewrite),
+                                                 CodeActionParams (..), Command,
+                                                 List (..), Method (..),
+                                                 SMethod (..), TextEdit (..),
+                                                 WorkspaceEdit (WorkspaceEdit),
+                                                 fromNormalizedUri,
+                                                 normalizedFilePathToUri,
+                                                 type (|?) (InR))
+import qualified Language.LSP.Types.Lens        as L
+
+data Log
+  = LogShake Shake.Log
+  | LogCollectedRecordSelectors [RecordSelectors]
+  | LogRenderedRecordSelectors [ConvertedRecordSelector]
+
+instance Pretty Log where
+  pretty = \case
+    LogShake shakeLog -> pretty shakeLog
+    LogCollectedRecordSelectors recs -> "Collected record selectors:" <+> pretty recs
+    LogRenderedRecordSelectors recs -> "Rendered record selectors:" <+> pretty recs
+
+data CollectRecordSelectors = CollectRecordSelectors
+                    deriving (Eq, Show, Generic)
+
+instance Hashable CollectRecordSelectors
+instance NFData CollectRecordSelectors
+
+data CollectConvertedRecordSelectorsResult = CCRSR
+  { recordInfos       :: RangeMap ConvertedRecordSelector
+  , enabledExtensions :: [GhcExtension]
+  }
+  deriving (Generic)
+
+instance NFData CollectConvertedRecordSelectorsResult
+
+instance Show CollectConvertedRecordSelectorsResult where
+  show _ = "<CollectRecordsResult>"
+
+type instance RuleResult CollectRecordSelectors = CollectConvertedRecordSelectorsResult
+
+-- `Extension` is wrapped so that we can provide an `NFData` instance
+-- (without resorting to creating an orphan instance).
+newtype GhcExtension = GhcExtension { unExt :: Extension }
```

Comment:
Looking at it, is there a specific reason why we don't just create an orphan? We already have a place for them, with plenty of orphans specifically for NFData. It seems like better something done in one central file, rather than every plugin define their own NFData instances.

Reply:
There wasn't a deep reason for why I went with the newtype solution, other than:
1)  adhering to the "orphans bad" idea
2) that's how it was (and still is) done in the plugin that I took as a reference for the general plugin structure (i.e. `hls-alternate-number-format-plugin`)

Avoiding orphan instances personally give me a peace of mind, but I agree that it shouldn't be a problem in this case. As long as we don't keep duplicating this code, I am happy with either solution (moving to a central place vs. orphan instance).

Reply:
I ended up adding an orphan instance for `Extension` (as well as `LHsExpr` for this plugin) and changing the explicit fields and this plugin to both use it.